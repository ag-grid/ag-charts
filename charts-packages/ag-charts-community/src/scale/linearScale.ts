import ContinuousScale from "./continuousScale";
import ticks, { tickIncrement } from "../util/ticks";
import { tickFormat } from "../util/numberFormat";

/**
 * Maps continuous domain to a continuous range.
 */
export class LinearScale extends ContinuousScale {
    readonly type = 'linear';

    ticks(count = 10) {
        const d = this._domain;
        return ticks(d[0], d[d.length - 1], count);
    }

    /**
     * Extends the domain so that it starts and ends on nice round values.
     * @param count Tick count.
     */
    nice(count = 10) {
        const d = this.domain;
        let i0 = 0;
        let i1 = d.length - 1;
        let start = d[i0];
        let stop = d[i1];
        let step;

        if (stop < start) {
            step = start;
            start = stop;
            stop = step;

            step = i0;
            i0 = i1;
            i1 = step;
        }

        step = tickIncrement(start, stop, count);

        if (step > 0) {
            start = Math.floor(start / step) * step;
            stop = Math.ceil(stop / step) * step;
            step = tickIncrement(start, stop, count);
        } else if (step < 0) {
            start = Math.ceil(start * step) / step;
            stop = Math.floor(stop * step) / step;
            step = tickIncrement(start, stop, count);
        }

        if (step > 0) {
            d[i0] = Math.floor(start / step) * step;
            d[i1] = Math.ceil(stop / step) * step;
            this.domain = d;
        } else if (step < 0) {
            d[i0] = Math.ceil(start * step) / step;
            d[i1] = Math.floor(stop * step) / step;
            this.domain = d;
        }
    }

    tickFormat(count?: number, specifier?: string) {
        const d = this.domain;
        return tickFormat(d[0], d[d.length - 1], count == undefined ? 10 : count, specifier);
    }
}