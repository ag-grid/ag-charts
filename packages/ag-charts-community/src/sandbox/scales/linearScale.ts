import { convertDomain } from '../util/domain.util';
import { ContinuousScale } from './continuousScale';

export class LinearScale extends ContinuousScale<number, number> {
    convert(value: number, clamp?: boolean) {
        return convertDomain(value, this.domain, this.range, clamp);
    }

    invert(value: number, clamp?: boolean) {
        return convertDomain(value, this.range, this.domain, clamp);
    }
}
