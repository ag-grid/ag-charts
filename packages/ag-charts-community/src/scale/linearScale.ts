import { ContinuousScale } from './continuousScale';
import ticks, { range, tickStep, singleTickDomain } from '../util/ticks';
import { tickFormat } from '../util/numberFormat';

/**
 * Maps continuous domain to a continuous range.
 */
export class LinearScale extends ContinuousScale<number> {
    readonly type = 'linear';

    interval?: number;

    public constructor() {
        super([0, 1], [0, 1]);
    }

    toDomain(d: number): number {
        return d;
    }

    ticks() {
        const count = this.tickCount ?? ContinuousScale.defaultTickCount;
        if (!this.domain || this.domain.length < 2 || count < 1 || this.domain.some((d) => !isFinite(d))) {
            return [];
        }
        this.refresh();
        const [d0, d1] = this.getDomain();

        const { interval } = this;

        if (interval) {
            const step = Math.abs(interval);
            if (!this.isDenseInterval({ start: d0, stop: d1, interval: step })) {
                return range(d0, d1, step);
            }
        }

        return ticks(d0, d1, count, this.minTickCount, this.maxTickCount);
    }

    update() {
        if (!this.domain || this.domain.length < 2) {
            return;
        }
        if (this.nice) {
            this.updateNiceDomain();
        }
    }

    /**
     * Extends the domain so that it starts and ends on nice round values.
     * @param count Tick count.
     */
    protected updateNiceDomain() {
        const count = this.tickCount ?? ContinuousScale.defaultTickCount;
        let [start, stop] = this.domain;
        if (count < 1) {
            this.niceDomain = [start, stop];
            return;
        }

        if (count === 1) {
            this.niceDomain = singleTickDomain(start, stop);
            return;
        }

        for (let i = 0; i < 2; i++) {
            const step = this.interval ?? tickStep(start, stop, count, this.minTickCount, this.maxTickCount);
            if (step >= 1) {
                start = Math.floor(start / step) * step;
                stop = Math.ceil(stop / step) * step;
            } else {
                // Prevent floating point error
                const s = 1 / step;
                start = Math.floor(start * s) / s;
                stop = Math.ceil(stop * s) / s;
            }
        }
        this.niceDomain = [start, stop];
    }

    tickFormat({ ticks, specifier }: { ticks?: any[]; specifier?: string }) {
        return tickFormat(ticks || this.ticks(), specifier);
    }
}
