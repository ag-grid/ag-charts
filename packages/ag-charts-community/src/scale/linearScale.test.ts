import { LinearScale } from './linearScale';

describe('LinearScale', () => {
    test('domain', () => {
        const scale = new LinearScale();

        expect(scale.domain).toEqual([0, 1]);
        scale.domain = [5, 10];
        expect(scale.domain).toEqual([5, 10]);
    });

    test('range', () => {
        const scale = new LinearScale();

        expect(scale.range).toEqual([0, 1]);
        scale.range = [5, 10];
        expect(scale.range).toEqual([5, 10]);
    });

    test('convert linear', () => {
        const scale = new LinearScale();

        scale.domain = [-100, 100];
        scale.range = [0, 100];

        expect(scale.convert(0, { clampMode: 'clamped' })).toBe(50);

        expect(scale.convert(-100, { clampMode: 'clamped' })).toBe(0);
        expect(scale.convert(100, { clampMode: 'clamped' })).toBe(100);

        expect(scale.convert(-100, { clampMode: 'raw' })).toBe(0);
        expect(scale.convert(100, { clampMode: 'raw' })).toBe(100);
    });

    test('convert linear clamp', () => {
        const scale = new LinearScale();

        scale.domain = [-100, 100];
        scale.range = [0, 100];

        expect(scale.convert(-300, { clampMode: 'clamped' })).toBe(0);
        expect(scale.convert(300, { clampMode: 'clamped' })).toBe(100);

        expect(scale.convert(-300, { clampMode: 'raw' })).toBe(-100);
        expect(scale.convert(300, { clampMode: 'raw' })).toBe(200);
    });

    test('convert linear with zero width domain', () => {
        const scale = new LinearScale();

        scale.domain = [100, 100];
        scale.range = [0, 100];

        expect(scale.convert(100, { clampMode: 'clamped' })).toBe(50);
    });

    test('invert linear', () => {
        const scale = new LinearScale();

        scale.domain = [-100, 100];
        scale.range = [0, 100];

        expect(scale.invert(50)).toBe(0);
        expect(scale.invert(0)).toBe(-100);
        expect(scale.invert(75)).toBe(50);
    });

    test('invert linear clamp', () => {
        const scale = new LinearScale();

        scale.domain = [-100, 100];
        scale.range = [0, 100];

        expect(scale.invert(-50)).toBe(-100);
        expect(scale.invert(150)).toBe(100);
    });

    test('invert linear with zero length range', () => {
        const scale = new LinearScale();

        scale.domain = [0, 100];
        scale.range = [100, 100];

        expect(scale.invert(100)).toBe(50);
    });

    describe('should create ticks', () => {
        const CASES = [
            {
                interval: 0,
                domain: [-1, 1],
            },
            {
                interval: 1,
                domain: [0, 10],
            },
            {
                interval: 3,
                domain: [-3, 12],
            },
            {
                interval: 10.5,
                domain: [0, 102],
            },
            {
                interval: 133,
                domain: [0, 665],
            },
            {
                interval: -1,
                domain: [0, 10],
            },
            {
                interval: -1,
                domain: [-10, 0],
            },
            {
                interval: -7.5,
                domain: [-37.5, -7.5],
            },
            {
                interval: 0.1,
                domain: [0, 1],
            },
            {
                interval: 0.01,
                domain: [0.1, 0.2],
            },
            {
                interval: 0.005,
                domain: [0.01, 0.02],
            },
        ];

        it.each(CASES)(`for interval: $interval domain: $domain case`, ({ interval, domain }) => {
            const scale = new LinearScale();

            scale.range = [0, 600];
            scale.domain = domain;
            scale.interval = interval;

            expect(scale.ticks()).toMatchSnapshot();
        });

        it('should create ticks within range', () => {
            const scale = new LinearScale();
            scale.range = [0, 100];
            scale.domain = [0, 100];
            scale.interval = 30;

            expect(scale.ticks()).toEqual([0, 30, 60, 90]);
        });
    });

    test('scale.tickFormat', () => {
        {
            const scale = new LinearScale();
            scale.domain = [-50000000, 50000000];
            const f = scale.tickFormat({ specifier: '~s' });
            expect(f(43000000)).toBe('43M');
        }
        {
            const scale = new LinearScale();
            scale.domain = [-50000000, 50000000];
            const f = scale.tickFormat({ specifier: '~s' });
            expect(f(-43000000)).toBe('âˆ’43M');
        }
        {
            const scale = new LinearScale();
            scale.domain = [-50000000, 50000000];
            const f = scale.tickFormat({ specifier: '~s' });
            expect(f(43500000)).toBe('44M');
        }
        {
            const scale = new LinearScale();
            scale.domain = [35000000, 44000000];
            const f = scale.tickFormat({ specifier: '~s' });
            const expectedTicks = ['36M', '38M', '40M', '42M', '44M'];
            const actualTicks = scale.ticks().map((t) => f(t));
            expect(actualTicks).toEqual(expectedTicks);
        }
        {
            const scale = new LinearScale();
            scale.domain = [3500000, 4400000];
            const f = scale.tickFormat({ specifier: '~s' });
            const expectedTicks = ['3.6M', '3.8M', '4M', '4.2M', '4.4M'];
            const actualTicks = scale.ticks().map((t) => f(t));
            expect(actualTicks).toEqual(expectedTicks);
        }
        {
            const scale = new LinearScale();
            scale.domain = [0.0034, 0.0044];
            const f = scale.tickFormat({ specifier: '~s' });
            const expectedTicks = ['3.4m', '3.6m', '3.8m', '4m', '4.2m', '4.4m'];
            const actualTicks = scale.ticks().map((t) => f(t));
            expect(actualTicks).toEqual(expectedTicks);
        }
        {
            const scale = new LinearScale();
            scale.domain = [0.0034, 0.0044];
            const f = scale.tickFormat({ specifier: 'f' });
            const expectedTicks = ['0.0034', '0.0036', '0.0038', '0.0040', '0.0042', '0.0044'];
            const actualTicks = scale.ticks().map((t) => f(t));
            expect(actualTicks).toEqual(expectedTicks);
        }
        {
            const scale = new LinearScale();
            scale.domain = [34, 44];
            const f = scale.tickFormat({ specifier: 'f' });
            const expectedTicks = ['34', '36', '38', '40', '42', '44'];
            const actualTicks = scale.ticks().map((t) => f(t));
            expect(actualTicks).toEqual(expectedTicks);
        }
        {
            const scale = new LinearScale();
            scale.domain = [35, 36];
            const f = scale.tickFormat({ specifier: 'f' });
            const expectedTicks = ['35.0', '35.2', '35.4', '35.6', '35.8', '36.0'];
            const actualTicks = scale.ticks().map((t) => f(t));
            expect(actualTicks).toEqual(expectedTicks);
        }

        const singlePointFormat = (n: number, specifier: string) => {
            const scale = new LinearScale();
            const pad = Math.abs(n) * 0.01;
            scale.domain = [n - pad, n + pad];
            const format = scale.tickFormat({ specifier, ticks: [n] });
            return format(n);
        };

        expect(singlePointFormat(0.1234567890123456, ' ')).toEqual('0.123456789012');
        expect(singlePointFormat(67.7, 'd')).toEqual('68');
        expect(singlePointFormat(0.678, '.2p')).toEqual('68%');
        expect(singlePointFormat(123, 'f')).toEqual('123');
        expect(singlePointFormat(0.001234567890123456, 'f')).toEqual('0.00123457');
        expect(singlePointFormat(0.1234567890123456, 'f')).toEqual('0.123457');
        expect(singlePointFormat(1.234567890123, 'f')).toEqual('1.23457');
        expect(singlePointFormat(123.4567890123, 'f')).toEqual('123.457');
        expect(singlePointFormat(12345.67890123, 'f')).toEqual('12345.7');
        expect(singlePointFormat(1234567.890123, 'f')).toEqual('1234568');
        expect(singlePointFormat(1234.567890123, 'f')).toEqual('1234.57');
        expect(singlePointFormat(1234.567890123, ' ')).toEqual('1234.56789012');
    });
});
