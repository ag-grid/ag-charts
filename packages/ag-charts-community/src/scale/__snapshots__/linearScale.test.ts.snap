// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinearScale should create ticks for interval: -1 domain: [-10, 0] case 1`] = `
[
  -10,
  -9,
  -8,
  -7,
  -6,
  -5,
  -4,
  -3,
  -2,
  -1,
  0,
]
`;

exports[`LinearScale should create ticks for interval: -1 domain: [0, 10] case 1`] = `
[
  0,
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
  10,
]
`;

exports[`LinearScale should create ticks for interval: -7.5 domain: [-37.5, -7.5] case 1`] = `
[
  -37.5,
  -30,
  -22.5,
  -15,
  -7.5,
]
`;

exports[`LinearScale should create ticks for interval: 0 domain: [-1, 1] case 1`] = `
[
  -1,
  -0.5,
  0,
  0.5,
  1,
]
`;

exports[`LinearScale should create ticks for interval: 0.005 domain: [0.01, 0.02] case 1`] = `
[
  0.01,
  0.015,
  0.02,
]
`;

exports[`LinearScale should create ticks for interval: 0.01 domain: [0.1, 0.2] case 1`] = `
[
  0.1,
  0.11,
  0.12,
  0.13,
  0.14,
  0.15,
  0.16,
  0.17,
  0.18,
  0.19,
  0.2,
]
`;

exports[`LinearScale should create ticks for interval: 0.1 domain: [0, 1] case 1`] = `
[
  0,
  0.1,
  0.2,
  0.3,
  0.4,
  0.5,
  0.6,
  0.7,
  0.8,
  0.9,
  1,
]
`;

exports[`LinearScale should create ticks for interval: 1 domain: [0, 10] case 1`] = `
[
  0,
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
  10,
]
`;

exports[`LinearScale should create ticks for interval: 3 domain: [-3, 12] case 1`] = `
[
  -3,
  0,
  3,
  6,
  9,
  12,
]
`;

exports[`LinearScale should create ticks for interval: 10.5 domain: [0, 102] case 1`] = `
[
  0,
  10.5,
  21,
  31.5,
  42,
  52.5,
  63,
  73.5,
  84,
  94.5,
]
`;

exports[`LinearScale should create ticks for interval: 133 domain: [0, 665] case 1`] = `
[
  0,
  133,
  266,
  399,
  532,
  665,
]
`;
